internal/pkg/composable/providers/kubernetessecrets/expiration_cache.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package kubernetessecrets
+
+import (
+	"sync"
+	"time"
+)
+
+// expirationCache is a store that expires items after time.Now - secret.lastAccess > ttl (if ttl > 0) at Get or List.
+// expirationCache works with *cacheEntry, a pointer struct that wraps secret, instead of secret directly because map
+// structure in standard go library never removes the buckets from memory even after removing all the elements from it.
+// However, since *cacheEntry is a pointer it can be garbage collected when no longer referenced by the GC, such as
+// when deleted from the map. More importantly working with a pointer makes the entry in the map bucket, that doesn't
+// get deallocated, to utilise only 8 bytes on a 64-bit system.
+type expirationCache struct {
+	sync.RWMutex
```

Comment:
we sure we need RWMutex, this does not add any performance gain and if write is frequent it can even result in a bad performance.
i'm leaning more towards normal Mutex, or we should benchmark to see benefits of having RWMutex.

Reply:
the choice of RWMutex, was solely done on the expectation that this would more read-based as each secret value would be inserted once either during the refresh of the cache or during a fetch, and then everything will hit for the ttl. I mean having a cache should make us read-heavy otherwise on each write we would be also fetching from the API server, which kinda defeats the purpose? I can still switch to plain Mutex just let me know 🙂

Reply:
i leave it up to you as RWMutex vs Mutex is a bit controversial topic